fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/api/ContainmentIndex.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.api;
+
+import java.util.stream.Stream;
+
+import org.fcrepo.kernel.api.models.FedoraResource;
+
+/**
+ * An interface defining the encapsulates containment relationships.
+ *
+ * @author dbernstein
+ * @since 6.0.0
+ */
+public interface ContainmentIndex {
+
+    /**
+     * Return a stream of fedora identifiers contained by the specified fedora resource.
+     *
+     * @param tx The transaction.  If no transaction, null is okay.
+     * @param fedoraResource The containing fedora resource
+     * @return A stream of contained identifiers
```

Comment:
I know you haven't implemented this yet, but should we define the order here. Below you assume they are in greatest grandchild to child order (I think). Maybe we should state that here, so its clear when we implement.

Or the other way around, we can reverse it.

Reply:
I am assuming that the index  returns identifiers that are directly contained by the specified object.  Thus the need for recursive calls to the index.

Reply:
Maybe the index should allow for a `getAllDescents()` query?

Reply:
My feeling is that we can optimize that later.

Reply:
Is this a `util`?  Or do you imagine this will store state?

Reply:
(I'm wondering if this should be in a sub directory)

Reply:
This will remain as a placeholder until @peichman-umd moves the SearchIndex interface into place.